spd/scripts/compare_models.py

```diff
@@ -0,0 +1,393 @@
+            current_flat = current_rank_one.reshape(int(C_alive.item()), -1)
+            ref_flat = ref_rank_one.reshape(C, -1)
+
+            current_norm = F.normalize(current_flat, p=2, dim=1)
+            ref_norm = F.normalize(ref_flat, p=2, dim=1)
+
+            cosine_sim_matrix = einops.einsum(
+                current_norm, ref_norm, "C_alive d_in_d_out, C_ref d_in_d_out -> C_alive C_ref"
+            )
+            cosine_sim_matrix = cosine_sim_matrix.abs()
+
+            max_similarities = cosine_sim_matrix.max(dim=1).values
+            similarities[f"mean_max_abs_cosine_sim/{layer_name}"] = max_similarities.mean().item()
+            similarities[f"max_abs_cosine_sim_std/{layer_name}"] = max_similarities.std().item()
+            similarities[f"max_abs_cosine_sim_min/{layer_name}"] = max_similarities.min().item()
+            similarities[f"max_abs_cosine_sim_max/{layer_name}"] = max_similarities.max().item()
+
+        metric_names = [
+            "mean_max_abs_cosine_sim",
+            "max_abs_cosine_sim_std",
+            "max_abs_cosine_sim_min",
+            "max_abs_cosine_sim_max",
+        ]
+
+        for metric_name in metric_names:
+            values = [
+                similarities[f"{metric_name}/{layer_name}"]
+                for layer_name in self.current_model.components
+                if f"{metric_name}/{layer_name}" in similarities
+            ]
+            if values:
+                similarities[f"{metric_name}/all_layers"] = sum(values) / len(values)
+
+        return similarities
+
+    def run_comparison(
+        self, eval_iterator: Iterator[Any], n_steps: int | None = None
+    ) -> dict[str, float]:
+        """Run the full comparison pipeline."""
+        if n_steps is None:
+            n_steps = self.comparison_config.get("n_eval_steps", 5)
+        assert isinstance(n_steps, int)  # Ensure n_steps is an int for type checking
+
+        logger.info("Computing activation densities for current model...")
+        activation_densities = self.compute_activation_densities(
+            self.current_model, eval_iterator, n_steps
+        )
+
+        logger.info("Computing geometric similarities...")
+        similarities = self.compute_geometric_similarities(activation_densities)
+
+        return similarities
+
+
+def main():
+    """Main execution function."""
+    parser = argparse.ArgumentParser(description="Compare two SPD models for geometric similarity")
+    parser.add_argument(
+        "--config",
+        type=str,
+        default="spd/scripts/compare_models_config.yaml",
+        help="Path to configuration file",
+    )
+    args = parser.parse_args()
+
+    config = load_config(args.config, dict)
+    current_model_path = config["current_model_path"]
+    reference_model_path = config["reference_model_path"]
+    density_threshold = config.get("density_threshold", 0.0)
+    device = config.get("device", "auto")
+    output_dir = Path(config.get("output_dir", "./comparison_results"))
```

Comment:
I think we should write results in an "out" directory that is relative to this specific file (`Path(__file__) / "out"`). As per my earlier comment which says that this compare models should be in its own directory within scripts, the outputs would then go in an out directory that's specific to the compare_models experiments. As it stands, this will write results in directories relative to wherever the user invokes this script. This is going to put them in their git working directory if they run it from the repo somewhere, and might put them in different paths if they invoke the script from different places.